pgjdbc/src/test/java/org/postgresql/test/util/ByteStreamWriterTest.java

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright (c) 2017, PostgreSQL Global Development Group
+ * See the LICENSE file in the project root for more information.
+ */
+
+package org.postgresql.test.util;
+
+import static org.junit.Assert.assertArrayEquals;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+import org.postgresql.test.TestUtil;
+import org.postgresql.test.jdbc2.BaseTest4;
+import org.postgresql.util.ByteBufferByteStreamWriter;
+import org.postgresql.util.ByteStreamWriter;
+
+import org.junit.Test;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.Random;
+
+public class ByteStreamWriterTest extends BaseTest4 {
+
+  @Override
+  public void setUp() throws Exception {
+    super.setUp();
+    assumeByteaSupported();
+    TestUtil.createTempTable(con, "images", "img bytea");
+  }
+
+  private ByteBuffer testData(int size) {
+    ByteBuffer data = ByteBuffer.allocate(size);
+    Random random = new Random(31459);
+    while (data.remaining() > 0) {
+      data.putLong(random.nextLong());
+    }
+    data.rewind();
+    return data;
+  }
+
+  private void insertStream(ByteBuffer testData) throws Exception {
+    insertStream(testData, null);
+  }
+
+  private void insertStream(ByteBuffer testData, Integer lengthOverride) throws Exception {
+    insertStream(new TestByteBufferByteStreamWriter(testData, lengthOverride));
+  }
+
+  private void insertStream(ByteStreamWriter writer) throws Exception {
+    PreparedStatement updatePS = con.prepareStatement(TestUtil.insertSQL("images", "img", "?"));
+    try {
+      updatePS.setObject(1, writer);
```

Comment:
Technically speaking, JDBC has a standard API for this: `java.sql.PreparedStatement#setBinaryStream(int, java.io.InputStream, long)`

Have you considered something like `implements InputStream, ByteStreamWriter`?

Reply:
The issue with this interface is that it gives the caller no control over buffering strategy and makes cleanup harder. Passing in an `InputStream` means that the driver will always copy the bytes to a fixed-length buffer( currently 8k) and then copy them out, see `PgStream.sendStream()`. The caller then has to infer that it's safe to clean up the underlying resource by waiting for the expected number of bytes to be read.

In contrast the writer interface lets the caller write data straight to the BufferedOutputStream that wraps the socket. If it's writing more data than the buffered stream has then that buffer will get flushed and the data will be copied over by the underlying native `write` implementation in `SocketOutputStream` in a single go.

The caller also knows when the write has finished (since they're in control) and can clean up any off-heap resources immediately.

Reply:
Re-reading, I think that you're suggesting that the `ByteStreamWriter` should also implement `InputStream` as a means to pass it in. Is that right? The thing is, it _isn't_ an `InputStream` and the JDBC API has a perfectly valid method to pass proprietary objects through: `setObject()` :). Can you explain why you don't like that approach?